third_party/libwebrtc/webrtc/modules/remote_bitrate_estimator/nada_core.cc

```diff
@@ -9,15 +9,578 @@
+                                            	  // [GAMMA_MAX: dimensionless]
+constexpr float   kNADAParamPlrSmooth = 0.1;  	// Smoothing factor for EMA of packet loss ratio
+
+constexpr float    kNADAParamQthMs =50.;    	  // Delay threshold for invoking non-linear warping 
+												                        // [QTH: in ms]
+constexpr float    kNADAParamLambda =0.5;   	  // Scaling parameter in the exponent of non-linear warping
+												                        // [LAMBDA: dimensionless]
+constexpr float    kNADAParamPlrRef =0.01;  	  // Reference packet loss ratio [PLRREF: dimensionless]
+constexpr float    kNADAParamDLossMs =10.;  	  // Reference delay penalty for loss when packet loss ratio
+										                      		  // is at PLRREF [DLOSS: dimensionless]
+}
+
+NadaCore::NadaCore()
+    : delta_(kNADAParamDeltaMs),
+      nada_dfwd_(0.),
+ 	    nada_dq_(0.),  
+ 	    nada_rtt_(0),  
+ 	    nada_relrtt_(0), 
+ 	    nada_nloss_(0), 
+ 	    nada_plr_(0.), 
+ 	    nada_rrate_(0.), 
+      nada_x_curr_(kNADAParamXDefault),
+      nada_x_prev_(kNADAParamXDefault), 
+      nada_rmode_(NADA_RMODE_ACCELERATED_RAMPUP),  
+      nada_rate_in_bps_(kNADAParamRateBps),
+      nada_rmin_in_bps_(kNADAParamRminBps),
+      nada_rmax_in_bps_(kNADAParamRmaxBps)  {}
 
 NadaCore::~NadaCore() {}
 
-void NadaCore::TestFunction (const char * from) const {
-    printf("[%s]: TestFunction__TestFunction__TestFunction__TestFunction__TestFunction__TestFunction\n", from);
+int NadaCore::GetMinBitrate() const {
+
+	return nada_rmin_in_bps_; 
+}
+
+/* 
+ * Update delay/loss states per feedback interval 
+ */
+void NadaCore::UpdateOwdStats(int64_t now_ms, 
+                              int64_t dfwd) {
+
+    UpdateOwdHistory(now_ms, dfwd); 
+
+    float d_base = min_owd_history_.front().second; 
+    nada_dfwd_ = dfwd; 
+    nada_dq_ = dfwd - d_base; 
+}
+
+void NadaCore::UpdateRttStats(int64_t now_ms, 
+                              int64_t rtt) {
+
+    UpdateRttHistory(now_ms, rtt); 
+
+    int64_t rtt_base = min_rtt_history_.front().second; 
+    nada_rtt_ = rtt; 
+    nada_relrtt_ = rtt = rtt_base; 
+}
+
+/*
+ * Optional TODO:  
+ * a) Currently, the instantaneous packet loss ratio is calculated based on the 
+ *    the number of packets lost over the most recent feedback interval; instead,
+ *    one could calculate the packet loss ratio following the TFRC-style algorithm
+ *    as explained in RFC5348 (https://www.rfc-editor.org/rfc/rfc5348.html). 
+ * b) Currently, the expiration window for last observed loss is fixed to be
+ *    the same as the most recent feedback interval; instead, one could
+ *    follow the descriptions in the draft (Sec. 4.2) to implement an adaptive loss
+ *    expiration window (loss_exp) that self-scales with average packet loss 
+ *    interval (loss_int). The latter is estimated following Sec. 5.4 of the TFRC
+ *    RFC (https://www.rfc-editor.org/rfc/rfc5348#section-5.4)
```

Comment:
I would add that these have been shown to improve the algorithm's behavior in ns3 simulations, but on real tests we couldn't rather enough data (so far).